_posts/blog/2016-10-08-vector_drawables_optimisation.md

```diff
@@ -0,0 +1,188 @@
+tags: [Android, UI, Vector Drawables, perfmatters]
+image:
+date: 2016-10-08T00:39:55-04:00
+---
+
+
+While some mobile platforms have been supporting vector graphics for a while, Android only began doing this natively starting with API Level 21 and with the help of the Support Library 23.2.0 for pre-Lollipop devices. By replacing you PNG image resources with VectorDrawables, your APK size decreases considerably and most of all, your images look good, independent of the resolution of the device used.
+
+
+The system will try to redraw your Activity every 16ms so it can reach the <a href="https://www.youtube.com/watch?v=CaMTIgxCSqU">60fps</a> target. So this means that the duration of every `onDraw` method is extremely important.
+
+
+When used incorrectly, VectorDrawables can affect the performance of your app, because the drawing can take a long time. Here are three mistakes that you might be making, but can be solved easily, to ensure that you are really improving the performance of your app, with the help of VectorDrawables.
+
+
+## Understanding The Internals Of VectorDrawables
+
+
+Vector graphics use geometrical shapes to describe graphical elements. The vector graphics are rendered at runtime. The automatic rendering at pixel density gives smoothness to the graphics, regardless of the device capabilities. So your images won’t be downscaled or upscaled, looking stretched or pixelated but they will be always perfectly drawn for your screen size.
+Vector drawables allow the representation of images (e.g. icons, UI elements) based on XML vector graphics.
+
+
+<center>
+<picture>
+	<a href="/images/blog/vector_drawables/vector_drawables_vs_png.png"><img src="/images/blog/vector_drawables/vector_drawables_vs_png.png" alt="VectorDrawable vs PNG" width="350"></a>
+	<figcaption>Using a VectorDrawable vs using a small PNG image</figcaption>
+</picture>
+</center>
+
+
+Using vector data instead of raster image data, the number and the size of resources added to the project decreases, since now only one resource per resolution is needed, and therefore also the APK size.
+
+
+Although the XML file containing the vector drawable is usually smaller than the PNG version, the vector drawables come with a computational overhead at runtime, which may be an issue for more complex graphical elements. When vector drawables are drawn for the first time, a cached bitmap is created in order to optimize the re-drawing performance. This cache is re-used as long as the width and the height of the image that needs to be drawn is the same. If a VectorDrawable is used for multiple sizes, a new Bitmap will be created every time.
+
+
+Compared to raster images, drawing VectorDrawables will take more time for the first rendering. But then, as long as the size of the image doesn't change, the next drawings of VectorDrawables will take a similar amount of time,  just like drawing raster images.
```

Comment:
But then, as long as the size of the image doesn't change, the next drawings of VectorDrawables will take as much time as drawing raster images.